aip/0162.md

```diff
@@ -0,0 +1,229 @@
+  created: 2019-09-17
+permalink: /162
+redirect_from:
+  - /0162
+---
+
+# Resource Revisions
+
+Some APIs need to have resources with a revision history, where users can
+reason about the state of the resource over time. There are several reasons for
+this:
+
+- Users may want to be able to roll back to a previous revision, or diff
+  against a previous revision.
+- An API may create data which is derived in some way from a resource at a
+  given point in time. In these cases, it may be desirable to snapshot the
+  resource for reference later.
+
+## Guidance
+
+APIs **may** store a revision history for a resource if it is useful to users.
+
+APIs implementing resources with a revision history **must** provide a
+`revision_id` field on the resource:
+
+```proto
+message Book {
+  // The name of the book.
+  string name = 1;
+
+  // Other fields…
+
+  // The revision ID of the book.
+  // This is automatically changed whenever the contents of the book
+  // change in any way.
+  string revision_id = 99 [(google.api.field_behavior) = OUTPUT_ONLY];
+}
+```
+
+- The `revision_id` field **should** be a string and contain a short,
+  automatically-generated hash. A good rule of thumb is the last eight
+  characters of a UUID4.
+- The field **must** document when new revisions are created.
+
+### Referencing revisions
+
+When it is necessary to refer to a specific revision of a resource, APIs
+**must** use the following syntax: `{resource_name}@{revision_id}`. For
+example:
+
+    publishers/123/books/les-miserables@c7cfa2a8
+
+APIs **should** generally accept a resource reference at a particular revision
+in any place where they ordinarily accept the resource name, except in
+situations that mutate the resource.
+
+**Important:** APIs **must not** require a revision ID, and **must** default to
+the current revision if one is not provided, except in methods specifically
+dealing with the revision history (such as rollback) where failing to require
+it would not make sense.
+
+### Getting a revision
+
+APIs implementing resource revisions **should** accept a resource name with a
+revision ID in the standard `Get` method ([AIP-131][]):
+
+```proto
+message GetBookRequest {
+  // The name of the book.
+  //   Example: publishers/123/books/les-miserables
+  //
+  // In order to retrieve a previous revision of the book, also provide
+  // the revision ID.
+  //   Example: publishers/123/books/les-miserables@c7cfa2a8
+  string name = 1;
+}
+```
+
+If the user passes a revision ID that does not exist, the API **must** fail
+with a `NOT_FOUND` error.
+
+### Listing revisions
+
+APIs implementing resource revisions **should** provide a custom method for
+listing the revision history for a resource, with a structure similar to
+standard `List` methods ([AIP-132][]):
+
+```proto
+rpc ListBookRevisions(ListBookRevisionsRequest)
+    returns (ListBookRevisionsResponse) {
+  option (google.api.http) = {
+    get: "/v1/{name=publishers/*/books/*}:listRevisions"
```

Comment:
Another possibility is `/revisions` rather than `:listRevisions`.
I think a custom verb is preferable, but I was not sure.

Reply:
Once I added the section for deleting revisions, it became very clear that `:listRevisions` is correct. (Otherwise `DeleteBook` and `DeleteBookRevision` would have identical URIs.)